### DISCUSSION TITLE
Scheduled tasks

### DISCUSSION THREAD

**Discussion by disconsented** [pufferpanel/pufferpanel#997] [Ideas]
I have appetite to implement scheduled tasks but little appetite to design the interface behind them and want to crowd source _sensible_ ideas.

Personal design goals:

- User friendly interface
- End user accessible
- Reliable(?) (tasks will retry?)

I think we can handle most use cases by slapping on _something_ (not cron) on top of operations.

For a first implementation, modelling this in the database is easy likely a new table with `serverid, operatation, args`.

The biggest thing I'd like to see discussion on is how you model what the user see and what you use to specify _when_ it will run, I will strongly advocate away from anything cron styled because its unusable for non technical people.

**Comment by Romitou** [comment]
Personally, I think that using cron would be ideal. PufferPanel could use libraries such as https://github.com/robfig/cron to schedule the tasks that users would have saved themselves, so users would not have to use crontab directly. Anyway, how do you want to schedule the tasks?

**Reply by disconsented** [reply]
Anything better than writing cron strings by hand, but it looks like nepcore has appetite to stick a GUI in front of cron which is fine.

**Comment by nepcore** [comment]
cron based scheduling expressions, be they executed by the PufferPanel service or through cron itself, doesn't sound like a big issue to me, cron expressions aren't super complex so I think it should be possible to rig up a frontend component to do exactly that (API use in my opinion doesn't need to be as careful about users with little to no technical background as I doubt such users will make much direct use of the API due to that being a rather technical thing as well), though if there are other solid options available that are easier to configure (for users and software) I would tend towards those too

It would probably be a good idea to support running more than one operation on a single task (just saying in case that's not already the plan) as e.g. backup routines not rarely need more than one thing to happen (example minecraft: send command to console to stop saving world to disk, copy world, send another command to reenable saving to disk)

I'd imagine a tasks steps being similar to the operations already used in templates but have 2 such options, one for sending a command to console and one to run an operation as used in templates (that or just have an additional operation for sending to the console that's only available for these tasks, though I don't know if golangs type system can actually handle such a thing and have a gut feeling it probably can't, then again, I have very little experience with golang, so I might be completely wrong here)

**Reply by disconsented** [reply]
> cron based scheduling expressions, be they executed by the PufferPanel service or through cron itself, doesn't sound like a big issue to me, cron expressions aren't super complex so I think it should be possible to rig up a frontend component to do exactly that (API use in my opinion doesn't need to be as careful about users with little to no technical background as I doubt such users will make much direct use of the API due to that being a rather technical thing as well), though if there are other solid options available that are easier to configure (for users and software) I would tend towards those too

My concern is about how it looks to the end user, if you're confident  that you can rig up something more user-friendly than [this](https://crontab.guru/) then I am quite happy to use cron on the backend.

> It would probably be a good idea to support running more than one operation on a single task (just saying in case that's not already the plan) as e.g. backup routines not rarely need more than one thing to happen (example minecraft: send command to console to stop saving world to disk, copy world, send another command to reenable saving to disk)

Good idea, I didn't consider that initially however, how do we handle partial failures? Generally I would go for an early failure.

> I'd imagine a tasks steps being similar to the operations already used in templates but have 2 such options, one for sending a command to console and one to run an operation as used in templates (that or just have an additional operation for sending to the console that's only available for these tasks, though I don't know if golangs type system can actually handle such a thing and have a gut feeling it probably can't, then again, I have very little experience with golang, so I might be completely wrong here)

This isn't a _big_ problem, just need to create a few new operators to achieve this and store things as slices.

- Console input operator
- Sleep operator (?)
- Console read operator (?)

**Reply by nepcore** [reply]
> My concern is about how it looks to the end user, if you're confident that you can rig up something more user-friendly than this then I am quite happy to use cron on the backend.

I want to say that making a decent frontend component for it is going to be a challenge regardless of the underlying format and asking the user to input text is probably not exactly reliable, even if we were to assume we somehow manage to deal with translation and have "natural input" like `every other day at 10 am`, some users would probably still not understand it, and I assume (without having done extensive research) there are probably more readily available tools to handle cron expressions than other things  
[This](https://karoletrych.github.io/vue-cron-editor/) is probably close to what I'd try to go for as that seems to me like a relatively user friendly presentation

> Good idea, I didn't consider that initially however, how do we handle partial failures? Generally I would go for an early failure.

Early failure and maybe save if the last run of the task was successful or not to the database so the frontend can show an indicator for tasks that failed on their last run  
Also, email to all users with permissions on the server when a task fails (maybe limiting by either whatever permission is used for managing tasks or through a manually configurable list defaulting to everyone)?

> This isn't a big problem, just need to create a few new operators to achieve this and store things as slices.
>
> * Console input operator
> * Sleep operator (?)
> * Console read operator (?)

Requesting server start/stop might also be useful as I've seen a decent amount of request for automated restarts  
And I'd say only run the next operation once the previous one finished (still might want a sleep operation for sending commands to console and giving them a second to go through, not sure how necessary that is though)

**Reply by disconsented** [reply]
> Early failure and maybe save if the last run of the task was successful or not to the database so the frontend can show an indicator for tasks that failed on their last run
Also, email to all users with permissions on the server when a task fails (maybe limiting by either whatever permission is used for managing tasks or through a manually configurable list defaulting to everyone)?

Short term emails would suffice, would want to have proper notifications longer term.

> Requesting server start/stop might also be useful as I've seen a decent amount of request for automated restarts
And I'd say only run the next operation once the previous one finished (still might want a sleep operation for sending commands to console and giving them a second to go through, not sure how necessary that is though)

That's my thinking as well, sleep should handle most edge cases even if it is a bit dirty.

Waiting on console read may be a bit expensive however, and would need an upper limit placed on how much time to wait (10m?)

**Comment by disconsented** [comment]
Hmm, can probably store tasks as part of a servers `.json`.

**Comment by Malachiel87** [comment]
any progress on this?
